core/src/cluster_info_vote_listener.rs

```diff
@@ -539,19 +608,62 @@ impl ClusterInfoVoteListener {
                         continue;
                     }
 
-                    for &slot in vote.slots.iter() {
-                        if slot <= root {
+                    let root = root_bank.slot();
+                    let last_vote_hash = vote.hash;
+                    for slot in &vote.slots {
+                        // If slot is before the root, or so far ahead we don't have
+                        // stake information, then ignore it
+                        let epoch = root_bank.epoch_schedule().get_epoch(*slot);
+                        let epoch_stakes = root_bank.epoch_stakes(epoch);
+                        if *slot <= root || epoch_stakes.is_none() {
                             continue;
                         }
+                        let epoch_stakes = epoch_stakes.unwrap();
+                        let epoch_vote_accounts = Stakes::vote_accounts(epoch_stakes.stakes());
+                        let total_epoch_stake = epoch_stakes.total_stake();
+
+                        let unduplicated_pubkey = vote_tracker.keys.get_or_insert(&vote_pubkey);
+
+                        // The vote for the greatest slot in the stack of votes in a non-switching vote
+                        // qualifies for optimistic confirmation.
+                        let update_switch_info = if (slot == last_vote_slot) && !is_switch_vote {
+                            let stake = epoch_vote_accounts
+                                .get(&vote_pubkey)
+                                .map(|(stake, _)| *stake)
+                                .unwrap_or(0);
+                            Some((stake, last_vote_hash))
+                        } else {
+                            None
+                        };
+
+                        // If this vote for this slot qualifies for optimistic confirmation
+                        if let Some((stake, hash)) = update_switch_info {
+                            // Fast track processing of non-switch votes so that
+                            // notifications for optimistic confirmation can be sent
+                            // as soon as possible.
+                            if !switch_counted.contains(&(*slot, *unduplicated_pubkey))
+                                && Self::add_non_switch_vote(
+                                    vote_tracker,
+                                    *slot,
+                                    hash,
+                                    unduplicated_pubkey.clone(),
+                                    stake,
+                                    total_epoch_stake,
+                                )
+                            {
+                                switch_counted.insert((*slot, *unduplicated_pubkey));
+                                new_optimistic_confirmed_slots.push((*slot, last_vote_hash));
+                                // TODO: Notify subscribers about new optimistic confirmation
```

Comment:
@CriesofCarrots fyi, this is where I'm going to try and plumb another notification level for "optimistically confirmed"